File: Python/Algorithms/Sorting-Algorithms/insertion_sort.py
# function to sort list of integers in place
# flag ascending is used to determine the sorting type i.e. ascending or descending
# insertion sort inserts elements into the left part of the array such that the left part is always sorted

# time complexity = O(n^2), but better than selection since best case complexity is O(n)

def insertionSort(a, ascending=True):
	n = len(a)
	# ranging from 1 and not 0 because 
	# a[0] by itself is always sorted
	for i in range(1, n):
		key = a[i]
		j = i-1

		while j > -1:
			if ascending:
				if key < a[j]:
					a[j], a[j+1] = a[j+1], a[j]
				else:
					break
			else:
				# for descenfing sort just compare -1 * value of elements
				if -key < -a[j]:
					a[j], a[j+1] = a[j+1], a[j]
				else:
					break
			j -= 1
		a[j+1] = key


# driver code
if __name__ == "__main__":
	# input a list of space separated integers
	sort_order = input("Enter 'd' for descending order of sort (by default ascending order is selected): ")
	l = list(map(int, input("Enter list to sort: ").split(" ")))

	print("Before sorting:", l)
	if sort_order == 'd':
		insertionSort(l, False)
		print("Sorting in descending order...")
	else:
		insertionSort(l)
		print("Sorting in ascending order...")
	print("After sorting:", l)

File: Haskell/Algorithms/Sorting-Algorithms/InsertionSort.hs
-- # Insertion Sort in Haskell
-- This is an implementation of simple Insertion Sort in Haskell on any type that can be ordered.
-- Insertion Sort is a slow, but simple sorting algorithm that has time complexity O(n^2)

-- Insert element into list in order, returning sorted list
insert :: (Ord a) => a -> [a] -> [a]
insert e []             = [e]
insert e (x:xs)
    | e < x             = e : x : xs
    | otherwise         = x : insert e xs

-- Insertion sort on list, returning the list sorted
insertionSort :: (Ord a) => [a] -> [a]
insertionSort []        = []
insertionSort (x:xs)    = insert x (insertionSort xs)

main :: IO ()
main = do
    let intList = [1, 30, 7, 10, 4, 25]
    print $ insertionSort intList -- Returning [1,4,7,10,25,30]